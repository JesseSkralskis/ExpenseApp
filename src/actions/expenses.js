import uuid from 'uuid'

//ADD_EXPENSE
//use named export
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: "ADD_EXPENSE",
  expenses: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

//REMOVE_EXPENSE

export const removeExpense = ({ id } = {}) => {


  
  return {
 
    type: "REMOVE_EXPENSE",
    id
  };
};
//EDIT_EXPENSE

export const editExpense = ({ id }, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});
