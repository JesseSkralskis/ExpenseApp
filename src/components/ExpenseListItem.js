import React from 'react'
import { NavLink } from "react-router-dom";
import moment from 'moment';
import numeral from 'numeral';

 



//this takes the props from expenseList (destructerted) and placesd them as needed
const ExpenseListItem = ({  id, description, amount, createdAt }) => (
  <div>
    <NavLink to={`/edit/${id}`}>
      {" "}
      <h3>{description}</h3>
    </NavLink>
    <p>
      {numeral(amount/100).format('$0,0.00')}
      -
      {moment(createdAt).format('MMMM Do, YYYY ')}
    </p>
    
  </div>
);
   //when we add connect we get dispatch as a prop even if we do no mapToProps function
    export default ExpenseListItem;