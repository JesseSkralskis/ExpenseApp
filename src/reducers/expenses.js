
//Expenses Reducer

//to simplify default values we creat a variable that reps it

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      //use concat because it does not change the array
      //better yet use spread operator
      //meaning it will not effect the original
      return [...state, action.expenses];

    case "REMOVE_EXPENSE":
     
      return state.filter(({ id }) => id !== action.id);
    
    

    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
        
    
          return {
            //using spread operators for objects
            //this sais take the existing 
            //with the id we gave and overwriting the updates we passed in
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });

    default:
      return state;
  }
};

export default expensesReducer;