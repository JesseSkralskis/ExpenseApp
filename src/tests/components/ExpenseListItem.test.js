import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import  ExpenseListItem  from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test('should render an instance of expense', () => {
    //instead of naming individual props we can use spread operator to get all the props much easier
    const description = expenses[0].description;
    const amount = expenses[0].amount;
    const  createdAt = expenses[0].createdAt;
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);

    expect(toJSON(wrapper)).toMatchSnapshot();
})