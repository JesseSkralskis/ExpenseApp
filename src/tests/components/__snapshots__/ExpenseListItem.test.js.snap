// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an instance of expense 1`] = `
<div>
  <NavLink
    to="/edit/1"
  >
     
    <h3>
      im a boss
    </h3>
  </NavLink>
  <p>
    195
    -
    0
  </p>
</div>
`;
